### PR TITLE
Add early-termination optimization to attenuated mip

### PR DESCRIPTION
**Pull request by aganders3**
This idea originated from a [thread on image.sc about poor framerates for volume rendering](https://forum.image.sc/t/poor-3d-viewer-performance-in-napari/68103) napari. The idea is to exit the main raycasting loop early if there is no chance of finding a higher value based on the provided contrast limits. I believe this will not affect the final appearance at all and is only a performance optimization (and my testing supports this).

I don't have a realistic dataset where this makes a truly meaningful difference, but for now you can try this example (adapted from `examples/scene/volume.py`) with a huge cube of random data with/without this change and the performance difference should be apparent.

```python
import numpy as np

from vispy import app, scene

# Prepare canvas
canvas = scene.SceneCanvas(size=(800, 600), show=True)
canvas.measure_fps()

# Set up a viewbox to display the image with interactive pan/zoom
view = canvas.central_widget.add_view()

# Create the volume visual
volume = scene.visuals.Volume(
    np.random.random((1024, 1024, 1024)).astype(np.float32),
    parent=view.scene,
    attenuation=0.05,
    method="attenuated_mip",
)

view.camera = scene.cameras.TurntableCamera(
    parent=view.scene,
    fov=60,
    name='Turntable',
)


if __name__ == '__main__':
    app.run()
```

N.B. @brisvag

### CODE REVIEW

**Comment by jni** [comment]
@aganders3 what do you think of my comment on the imagesc thread that the contrast limits should actually apply to the *attenuated* value, rather than the original data? Basically, my idea is that the attenuation should affect *which* voxel gets rendered from the ray, and then the mapping of that voxel value to a color gets put through the colormap. Am I thinking about this wrong?

**Comment by aganders3** [comment]
> @aganders3 what do you think of my comment on the imagesc thread that the contrast limits should actually apply to the _attenuated_ value, rather than the original data? Basically, my idea is that the attenuation should affect _which_ voxel gets rendered from the ray, and then the mapping of that voxel value to a color gets put through the colormap. Am I thinking about this wrong?

Yes I want to look into this. I didn't want to hijack the image.sc thread, so was going to start a Zulip chat about it but just haven't made the time yet.

Have you tried it since the changes in https://github.com/vispy/vispy/pull/2417? I feel it is much more predictable since then but there's surely room for improvement. I still think it makes sense to scale/clamp when calculating the attenuation such that the attenuation is not dependent on the scale of your data (in particular when you have negative values this does not make sense) but I see where you're coming from as well. I'm going to try a few things.

**Comment by jni** [comment]
@aganders3 Interesting, thanks for the link! Totally missed that PR. I'm convinced that I have no idea what the right thing to do is, carry on! 😂

**Comment by aganders3** [comment]
> You could do a few buffer captures at different levels of attenuation and compare the result between the two implementations to actually make sure this is true. But in principle I agree that it should be the same.

Thanks for the idea - I added this to my demo script and I'm satisfied it's the same 🙂

> As far as I can tell the failures are caused by the new ipython/jupyter comm package being released and/or used before it was ready (NotImplementedError). The code has changed quite a bit in the last 24 hours or so: https://github.com/ipython/comm

Anything I can help with or should it just be re-run? It looks like the test used `comm=0.1.2` and now 0.1.3 is released and has a relevant fix: https://github.com/ipython/comm/pull/13

**Comment by djhoese** [comment]
Restarted failing job. Let's see what happens.

**Comment by aganders3** [comment]
Now that I think about it - the same optimization can be applied to non-attenuated MIP for any fragment that's going to be maxed out (`maxval >= clim.y`). This is probably less impactful, but may still be worth including. I guess it's already in place here if you set `attenuation = 0` to get a standard MIP, so that gives a relative simple way to roughly compare performance (with the overhead that attenuation is still calculated unnecessarily for each step).

**Comment by jni** [comment]
> This is probably less impactful,

I disagree, because MIP is the default rendering that most people use. I expect that the overhead of attenuation is actually significant and you should make this optimisation! (I thought that was already in place...)

**Comment by psobolewskiPhD** [comment]
Running the example takes considerable time before anything happens (with a 10Gb+ python process spawned), but the difference in performance with this PR is astounding.
Without I get max 4.9 fps, with 60 fps all the time.

**Comment by aganders3** [comment]
> I disagree, because MIP is the default rendering that most people use. I expect that the overhead of attenuation is actually significant and you should make this optimisation! (I thought that was already in place...)

I mean less impactful in that a MIP with reasonable contrast limits will probably not have too many saturated pixels, where the attenuated MIP might have many that reach attenuation. In any case I added this + the same for minip. I also added a flag to skip the fine-step refinement in these cases.

> Running the example takes considerable time before anything happens (with a 10Gb+ python process spawned), but the difference in performance with this PR is astounding.
Without I get max 4.9 fps, with 60 fps all the time.

Yes - sorry for the enormous example :p. The random noise likely shows the effect much more than real data, but it is a striking difference!

**Comment by aganders3** [comment]
Oops, 9e61444 was bad, but I think it's fixed in 3cb973a. With test data the buffer is still identical with these changes, but there could still be some difference with the depth as it's now sometimes returning on the _first_ value above/below the threshold along the ray. I don't think this would affect the output image but might be missing something.

**Comment by jni** [comment]
> there could still be some difference with the depth as it's now sometimes returning on the first value above/below the threshold along the ray.

I think this is actually what users would expect, to be honest?

**Comment by brisvag** [comment]
As far as I can tell this is ready to go in! @djhoese, should we get this in the next release?

File: vispy/visuals/volume.py [hunk 1]
```diff
@@ -367,16 +367,22 @@
     before_loop="""
         float maxval = u_mip_cutoff; // The maximum encountered value
         int maxi = -1;  // Where the maximum value was encountered
+        bool refine = true; // Whether to refine the frag with smaller steps
```
Comment (brisvag) [inline vispy/visuals/volume.py:370]: Actually, why is this needed? It could still be that the actual best value is *nearby*, so I don't see why we would skip refiniment in these cases. Sure, it improves performance, but then you have a slightly different spatial resolution along the ray cast depending on whether the value was marked as "higher than clim" or not, which is a bit inconsistent.

**Review by brisvag — CHANGES_REQUESTED** [review]
Sorry for the last minute doubt; I don't think it's warranted to add that `refine` flag, unless I misunderstand how it works?

File: vispy/visuals/volume.py (hunk 1 above)
Comment (aganders3) [reply vispy/visuals/volume.py:370]: I think you're right. My thinking was if the pixel is going to be pure white (or whatever max val) we can stop all additional work, but this will potentially get the wrong depth slightly. My understanding is a little thin here but I suppose that could be a problem with multiple visuals? In any case this was a pretty small improvement if any so I am happy to remove it.

Comment (brisvag) [reply vispy/visuals/volume.py:370]: Yeah it's probably minimal, so I would err on the side of not having a different behaviour from normal :+1:

**Comment by brisvag** [comment]
@djhoese feel free to merge if this is good for you!

**Comment by djhoese** [comment]
LGTM. Thanks @aganders3!
